### PR TITLE
[BACKPORT] Fix incorrect indexing inte Merge parameters in -rock-fold-broadcast

### LINKED ISSUE #1417: [Issue]: Assertion `Index < Length && "Invalid index!"' failed
### Problem Description

Build MIGraphX `develop` branch + https://github.com/ROCm/AMDMIGraphX/pull/2782 
I was using a ROCm 6.0.2 docker
I ran the migraphx-driver and hit the error listed in the "Additional Information" section of this issue.

### Operating System

20.04.6

### CPU

AMD EPYC 7702 64-Core Processor

### GPU

AMD Instinct MI100

### ROCm Version

ROCm 6.0.0

### ROCm Component

rocMLIR

### Steps to Reproduce

_No response_

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

_No response_

### Additional Information

Trace collected from MIGRAPHX_TRACE_MLIR=1 /workspace/AMDMIGraphX/build/bin/driver perf /new-saved-models/tiny_llama/decoder_model.onnx --fill1 input_ids attention_mask --inpu
t-dim @input_ids 1 256 @attention_mask 1 256

[mlir_error_tinylama.txt](https://github.com/ROCm/rocMLIR/files/14316147/mlir_error_tinylama.txt)

### PR DESCRIPTION
**Pull request by krzysz00**
Original PR: #1419 

Fixes #1417

FoldBroadcast would crash when processing Merge operations like <Merge{4, 1} ["x"] at [1] -> ["a", "b"] at [1, 2]>, as it would index the merge parameters with 1 and 2 and not 0 and 1, causing out of bounds array accesses and thus assertion failures.

This commit resolves that issue.

In addition, while looking around nearby, I identifidied that the handling for Broadcast{} operations that handled multiple dimensions (like, say,
<Broadcast{1, 1} ["a", "b"] at [0, 1] -> ["a", "b"] at [0, 1]> ) was incorrect because the worklist update logic always checked the last underlying dimension for Broadcast{}. This commit fixes that issue.

### CODE REVIEW

**Review by manupak — APPROVED** [review]
approving the backport....

**Review by jerryyin — APPROVED** [review]
